File: writebook/app/models/p2u/parser.rb
require "yaml"

module P2u
  # Turns a manifest source string into a P2u::Manifest. Supports three surfaces
  # that compile to the same AST:
  #
  #   JSON      — machine / tool-call friendly
  #   YAML      — canonical, human and agent friendly
  #   Markdown  — front matter for deck meta, `---` between slides
  class Parser
    MARKDOWN_SEPARATOR = /\n[ \t]*---[ \t]*\n/
    FORMATS = %w[json yaml yml markdown md].freeze

    def self.parse(source, format: nil)
      new(source, format: format).parse
    end

    def initialize(source, format: nil)
      @source = source.is_a?(String) ? source : source.to_json
      @format = format&.to_s&.strip&.downcase.presence
    end

    def parse
      raise ParseError, "Manifest source is empty" if @source.strip.empty?

      format = @format || detect_format
      case format
      when "json" then parse_json
      when "yaml", "yml" then parse_yaml
      when "markdown", "md" then parse_markdown
      else raise ParseError, "Unsupported format: #{format.inspect} (expected one of #{FORMATS.join(', ')})"
      end
    rescue ParseError
      raise
    rescue StandardError => e
      raise ParseError, "Could not parse #{format} manifest: #{e.message}"
    end

    private
      def detect_format
        stripped = @source.lstrip
        return "json" if stripped.start_with?("{", "[")
        return "markdown" if markdown_front_matter?(stripped)
        return "markdown" if @source.match?(MARKDOWN_SEPARATOR) && !yaml_like?(stripped)

        "yaml"
      end

      # A Markdown document has front matter delimited by a second `---`; a
      # plain YAML document only has the leading document marker.
      def markdown_front_matter?(text)
        text.start_with?("---") && text.match?(/\A---[ \t]*\r?\n.*?\r?\n---[ \t]*(\r?\n|\z)/m)
      end

      def yaml_like?(text)
        text.match?(/\A(p2u|version|deck|slides)\s*:/)
      end

      def parse_json
        Manifest.new(JSON.parse(@source))
      end

      def parse_yaml
        Manifest.new(YAML.safe_load(@source, permitted_classes: [ Date, Time ], aliases: true) || {})
      end

      def parse_markdown
        text = @source.dup
        meta = {}

        if text.start_with?("---")
          rest = text.sub(/\A---\s*\n/, "")
          if rest =~ /\n---\s*\n/
            front, body = rest.split(/\n---\s*\n/, 2)
            parsed = YAML.safe_load(front, permitted_classes: [ Date, Time ], aliases: true)
            meta = parsed.is_a?(Hash) ? parsed : {}
            text = body.to_s
          end
        end

        deck = meta["deck"].is_a?(Hash) ? meta["deck"] : meta.except("p2u", "version", "slides")
        slides = text.split(MARKDOWN_SEPARATOR).filter_map { |chunk| markdown_slide(chunk) }

        Manifest.new({ "p2u" => meta["p2u"] || meta["version"] || VERSION, "deck" => deck, "slides" => slides })
      end

      def markdown_slide(chunk)
        text = chunk.to_s.strip
        return if text.blank?

        if text.start_with?("# ")
          heading, body = split_heading(text)
          { "layout" => "section", "title" => heading, "body" => body.presence }.compact
        elsif text.start_with?("## ")
          heading, body = split_heading(text)
          { "layout" => "content", "title" => heading, "body" => body.presence }.compact
        else
          { "layout" => "content", "body" => text }
        end
      end

      def split_heading(text)
        lines = text.lines
        heading = lines.first.to_s.sub(/\A\#{1,6}\s*/, "").strip
        body = lines.drop(1).join.strip
        [ heading, body ]
      end
  end
end

File: cloudflare/src/compiler/parser.ts
// Turns a manifest source string into a Manifest. Supports three surfaces that
// compile to the same AST:
//
//   JSON      — machine / tool-call friendly
//   YAML      — canonical, human and agent friendly
//   Markdown  — front matter for deck meta, `---` between slides
//
// Port of p2u/parser.rb.

import { parse as parseYaml } from "yaml";
import { Manifest } from "./manifest";
import { ParseError } from "./diagnostic";
import { isBlank, presence } from "./util";
import { VERSION } from "./version";

export type Format = "json" | "yaml" | "markdown";

const MARKDOWN_SEPARATOR = /\n[ \t]*---[ \t]*\n/;
const FORMATS = ["json", "yaml", "yml", "markdown", "md"];

export class Parser {
  private readonly source: string;
  private readonly format?: string;

  constructor(source: unknown, format?: string | null) {
    this.source = typeof source === "string" ? source : JSON.stringify(source);
    this.format = presence(format)?.toLowerCase();
  }

  static parse(source: unknown, format?: string | null): Manifest {
    return new Parser(source, format).parse();
  }

  parse(): Manifest {
    if (isBlank(this.source)) throw new ParseError("Manifest source is empty");

    const format = this.format ?? detectFormat(this.source);
    try {
      switch (format) {
        case "json":
          return parseJson(this.source);
        case "yaml":
        case "yml":
          return parseYamlDocument(this.source);
        case "markdown":
        case "md":
          return parseMarkdown(this.source);
        default:
          throw new ParseError(`Unsupported format: ${JSON.stringify(format)} (expected one of ${FORMATS.join(", ")})`);
      }
    } catch (error) {
      if (error instanceof ParseError) throw error;
      throw new ParseError(`Could not parse ${format} manifest: ${(error as Error).message}`);
    }
  }
}

function detectFormat(source: string): Format {
  const stripped = source.replace(/^\s+/, "");
  if (stripped.startsWith("{") || stripped.startsWith("[")) return "json";
  if (markdownFrontMatter(stripped)) return "markdown";
  if (MARKDOWN_SEPARATOR.test(source) && !yamlLike(stripped)) return "markdown";
  return "yaml";
}

/** A Markdown document has front matter delimited by a second `---`. */
function markdownFrontMatter(text: string): boolean {
  return text.startsWith("---") && /^---[ \t]*\r?\n[\s\S]*?\r?\n---[ \t]*(\r?\n|$)/.test(text);
}

function yamlLike(text: string): boolean {
  return /^(p2u|version|deck|slides)\s*:/.test(text);
}

function parseJson(source: string): Manifest {
  return new Manifest(JSON.parse(source));
}

function parseYamlDocument(source: string): Manifest {
  return new Manifest(parseYaml(source) ?? {});
}

function parseMarkdown(source: string): Manifest {
  let text = source;
  let meta: Record<string, any> = {};

  if (text.startsWith("---")) {
    const rest = text.replace(/^---\s*\n/, "");
    const match = rest.match(/\n---\s*\n/);
    if (match && match.index !== undefined) {
      const front = rest.slice(0, match.index);
      const body = rest.slice(match.index + match[0].length);
      const parsed = parseYaml(front);
      meta = parsed && typeof parsed === "object" ? parsed : {};
      text = body;
    }
  }

  const deck = meta.deck && typeof meta.deck === "object" ? meta.deck : omit(meta, ["p2u", "version", "slides"]);
  const slides = text
    .split(MARKDOWN_SEPARATOR)
    .map((chunk) => markdownSlide(chunk))
    .filter((slide): slide is Record<string, unknown> => slide !== null);

  return new Manifest({ p2u: meta.p2u ?? meta.version ?? VERSION, deck, slides });
}

function markdownSlide(chunk: string): Record<string, unknown> | null {
  const text = chunk.trim();
  if (isBlank(text)) return null;

  if (text.startsWith("# ")) {
    const { heading, body } = splitHeading(text);
    return compact({ layout: "section", title: heading, body: presence(body) });
  }
  if (text.startsWith("## ")) {
    const { heading, body } = splitHeading(text);
    return compact({ layout: "content", title: heading, body: presence(body) });
  }
  return { layout: "content", body: text };
}

function splitHeading(text: string): { heading: string; body: string } {
  const lines = text.split("\n");
  const heading = (lines[0] ?? "").replace(/^#{1,6}\s*/, "").trim();
  const body = lines.slice(1).join("\n").trim();
  return { heading, body };
}

function omit(source: Record<string, any>, keys: string[]): Record<string, any> {
  const out: Record<string, any> = {};
  for (const [key, value] of Object.entries(source)) {
    if (!keys.includes(key)) out[key] = value;
  }
  return out;
}

function compact(source: Record<string, unknown>): Record<string, unknown> {
  const out: Record<string, unknown> = {};
  for (const [key, value] of Object.entries(source)) {
    if (value !== undefined && value !== null) out[key] = value;
  }
  return out;
}